src/VisualStudio/Core/Def/Telemetry/AggregatingTelemetryLogManager.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.Telemetry;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Telemetry
+{
+    /// <summary>
+    /// Manages creation and obtaining aggregated telemetry logs. Also, notifies logs to
+    /// send aggregated events every 30 minutes or upon disposal.
+    /// </summary>
+    internal sealed class AggregatingTelemetryLogManager : IDisposable
+    {
+        private readonly TelemetrySession _session;
+        private readonly Dictionary<FunctionId, AggregatingTelemetryLog> _aggregatingLogs;
+        private readonly object _lock;
+        private const int BatchedTelemetryCollectionPeriodInSeconds = 60 * 30;
```

Comment:
doc.  and/or use TimeSpan.  move to top.  Does this need to be const?  can this be a feature-flag so we can adjust this dynamically if we feel there's a problem?  e.g. telemetry folks say "please lower this to every few hours"

Reply:
I feel like a feature flag for this might be a bit overkill. Even if we increased this timeout to forever, we would still log the aggregated events on VS close. Feel free to push back and I can pursue the change, it just doesn't seem like we're collecting anywhere near enough aggregated events for a purge every 30 minutes to be problematic.